File: leetcode/#179.py
# question --> https://leetcode.com/problems/largest-number/

import functools
class Solution:
    def compare(self, string1: str, string2: str) -> int:
        if string1 + string2 >= string2 + string1:
            return -1
        else:
            return 1
    
    def largestNumber(self, nums: List[int]) -> str:
        nums = [str(num) for num in nums]
        nums = sorted(nums, key = functools.cmp_to_key(self.compare))

        return str(int(''.join(nums)))
        
        # TC: O(NlogN); SC: O(1)
        
# end 

File: leetcode/#179.js
// question --> https://leetcode.com/problems/largest-number/

/**
 * @param {number[]} nums
 * @return {string}
 */
var largestNumber = function(nums) {
    const ans = nums.map(String).sort((a, b) => a.concat(b) >= b.concat(a) ? -1 : 1).join('');
    return ans.charAt(0) === '0' ? '0' : ans; 
};

// TC: O(NlogN); SC: O(N)

// end 